Moving the shared enums to gfx_traits; r=jdm

Wafflespeanut opened the pull request:
This is just a draft for showing my initial commit for [issue #5698](https://github.com/servo/servo/issues/5698). I'll continue updating the commit until the build succeeds :) /cc @jdm

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/6685)

<!-- Reviewable:end -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @jdm (or someone else) soon.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

jdm commented on components/gfx_traits/paint_task.rs:
```diff
@@ -0,0 +1,69 @@
+//! The task that handles all painting.
+
+use azure::azure_hl::Color;
+use display_list::StackingContext;
```
Why is this here?

jdm commented on components/gfx/display_list/mod.rs:
```diff
@@ -17,6 +17,7 @@
 #![deny(unsafe_code)]
 
 use display_list::optimizer::DisplayListOptimizer;
+use gfx_traits::display_list::*;
```
Why is this change necessary?

jdm commented on components/gfx/paint_task.rs:
```diff
@@ -39,60 +39,34 @@ use util::task::spawn_named_with_send_on_failure;
-    pub epoch: Epoch,
-    pub layer_kind: LayerKind,
-}
+use gfx_traits::paint_task::{CompositorPaintMsg, PaintTask};
```
PaintTask doesn't exist yet, right?

jdm commented on components/gfx/platform/freetype/font_template.rs:
```diff
@@ -5,16 +5,7 @@
-    pub bytes: Vec<u8>,
-    pub identifier: Atom,
-}
+use gfx_traits::platform::font_template::freetype::FontTemplateData;
```
Why this change?

jdm commented on components/gfx/platform/macos/font_template.rs:
```diff
@@ -13,24 +13,7 @@ use std::borrow::ToOwned;
-    pub identifier: Atom,
-    pub font_data: Option<Vec<u8>>
-}
+use gfx_traits::platform::font_template::macos::FontTemplateData;
```
Why this change?

jdm commented on components/gfx/text/glyph.rs:
```diff
@@ -3,6 +3,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use euclid::point::Point2D;
+use gfx_traits::text::glyph::*;
```
Why this change?

jdm commented on components/gfx/text/text_run.rs:
```diff
@@ -4,6 +4,7 @@
 
 use font::{Font, FontHandleMethods, FontMetrics, IS_WHITESPACE_SHAPING_FLAG, RunMetrics};
 use font::{ShapingOptions};
+use gfx_traits::text::text_run::*;
```
Why this change?

jdm commented:
This is a good start, but I think there's a lot of code being moved around here this isn't necessary, and I suspect it stems from the unnecessary inclusion of `StackingContext` in `gfx_traits/paint_task.rs`.

Wafflespeanut commented:
@jdm Initially, I had moved the `StackingContext` inside `gfx_traits` because `Msg` demanded it. Now that we've split the `Msg` enum, it isn't of any use anymore (and, I forgot about that). So yeah, you're totally right. Will make this work today :)

Wafflespeanut commented:
@jdm r? (and, I shall proceed to PaintTask and font_cache_task) :)

jdm commented:
Much better!

jdm commented:
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 3 of 20 files at r1, 20 of 20 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/gfx/paint_task.rs, line 262 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6685#-JupmHL85nI3v0WgYPAt)** ([raw file](https://github.com/servo/servo/blob/49d0e11b21f5437d2947fc691aaf1102875d03fd/components/gfx/paint_task.rs#L262)):</sup>
Move this to a new line.

---

<sup>**[components/gfx_traits/lib.rs, line 4 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6685#-Jupn3UjInjOlYNmbU4_)** ([raw file](https://github.com/servo/servo/blob/5caf1301692c5b53c300ef13a83a8f2e8c1519bd/components/gfx_traits/lib.rs#L4)):</sup>
Whoops!

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6685)
<!-- Sent from Reviewable.io -->

Wafflespeanut commented:
@jdm Missed that. Fixed now :)

jdm commented:
@bors-servo: r+
-S-awaiting-review +S-awaiting-merge   

---

Reviewed 2 of 2 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6685)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 4b8bea5 has been approved by `jdm`

<!-- @bors-servo r=jdm 4b8bea52108ecde804df985e35d25264aeacda43 -->

jdm commented:
@bors-servo: r-

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 7758b6c has been approved by `jdm`

<!-- @bors-servo r=jdm 7758b6c1ed116181bdce9317af7bd4eb5d5b4762 -->

bors-servo commented:
:hourglass: Testing commit 7758b6c with merge 3bd322d...

jdm commented:
Missing appropriate changes to the various Cargo.lock files.

Wafflespeanut commented:
@jdm Build succeeds locally. r?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6679) made this pull request unmergeable. Please resolve the merge conflicts.

Wafflespeanut commented:
@bors-servo Ohhhhh youuuu!!! (I gotta reset, stash & force push again) -_-

jdm commented:
You can also use the power of rebase: http://rypress.com/tutorials/git/rebasing

Wafflespeanut commented:
@jdm Ooooh, nice! Thanks. (sadly, /me has already done the exhaustive merging -_-)

Wafflespeanut commented:
@jdm Ooooh, nice! Thanks. (sadly, /me has already done the exhaustive merging -_-). Anyways, r?

pcwalton commented:
Unfortunately, I think this should wait on e10s. It'll be a lot of churn otherwise.

Wafflespeanut commented:
@pcwalton Oh, take your time. Ping me when you're done :)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6742) made this pull request unmergeable. Please resolve the merge conflicts.

metajack commented:
@pcwalton Which PR should this be blocked on? Or did you mean all e10s PRs?

Wafflespeanut commented:
Looks like this PR hasn't got much work anymore (since @pcwalton has done most of it). All there's left is to just move the public `ChromeToPaintMsg` enum to `gfx_traits`.

Wafflespeanut commented:
Alright, this doesn't work yet! `msg` depends on `canvas_traits` which depends on `gfx_traits` which further depends on `msg`. So, cyclic dependency!

Wafflespeanut commented:
Alright, this doesn't work yet! There's a cyclic dependency! Here's how they depend on one another:

- `msg::constellation_msg` --> `canvas_traits::CanvasMsg`
- `canvas_traits` --> `gfx_traits::color`
- `gfx_traits::paint_task` --> `msg::{constellation_msg, compositor_msg}`

And, none of the enums can be moved from one to another, because they appear as a *core* enum in all these crates (i.e., they're being used by a lot of other modules in other crates).

So, I can't proceed @jdm :/

jdm commented:
The only solution I've come up with is splitting the giant constellation_msg::Msg enum into FromScript and FromCompositor enums, and moving the FromScript one to `script_traits`. This would mean adding a second receiver to the constellation, and making it select over both of them, etc., but it would remove the `msg -> canvas_traits` dependency. Fun times!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6970) made this pull request unmergeable. Please resolve the merge conflicts.

Wafflespeanut commented:
@jdm Once we split the `Msg` and add a new receiver to `Constellation`, it'd then require a sender, for which we'll need to create something (say 'ScriptConstellationChan') similar to the `ConstellationChan` wrapper struct (for utilizing `MsgFromScript`). Now, both the wrapper structs should be added to `Constellation`.

But, the [`start`](https://mxr.mozilla.org/servo/source/components/compositing/constellation.rs#231) method returns a `ConstellationChan`, which doesn't include the `ScriptConstellationChan`. How should I overcome this problem?

And, it'd make a considerable difference in my view if I understand [this `impl`](https://mxr.mozilla.org/servo/source/components/compositing/constellation.rs#221). I can't even see where this `start` method is used :/

jdm commented:
http://mxr.mozilla.org/servo/source/components/servo/lib.rs#169 is where Constellation::start gets used.

jdm commented:
I'm imagining the new channel being stored inside Constellation and being passed to new pipelines that are created.

Wafflespeanut commented:
@jdm Awesome! Thanks :)

Wafflespeanut commented:
@jdm Thanks. Now, this gets interesting - [canvasrenderingcontext2d.rs](https://mxr.mozilla.org/servo/source/components/script/dom/canvasrenderingcontext2d.rs#125) and [webglrenderingcontext.rs](https://mxr.mozilla.org/servo/source/components/script/dom/webglrenderingcontext.rs#71) are the only files that are gonna be using the variants from `ScriptConstellationChan`, but they seem to get the thing from `GlobalRef`, which means both the `Worker` and the `Window` must have another field for `ScriptConstellationChan`, right? So, this is gonna get messy.

Once that's done, I could go for a `global.constellation_chan(args)` where `args` would typically be a `bool` asking whether to supply a `CompositorConstellationChan` or `ScriptConstellationChan` or I could also go for the good old procedure of adding another method for `script_constellation_chan` to the global. Either way, this is gonna get complicated, as it's probably gonna modify a great deal of files. Any ideas?

jdm commented:
@Wafflespeanut I'm pretty sure that Worker and Window could replace their constellation chan with the new ScriptConstellationChan, if the name actually describes its purpose :)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7238) made this pull request unmergeable. Please resolve the merge conflicts.

Wafflespeanut commented:
Meh...

Wafflespeanut commented:
@jdm Now, there's another iceberg. `ScriptTask` creates a `NewLayoutTaskInfo` using its own [`constellation_chan`](https://mxr.mozilla.org/servo/source/components/script/script_task.rs#1092) (which is now a `ScriptConstellationChan`). But, it can't have that, because it's then used to create a [`LayoutTask`](https://mxr.mozilla.org/servo/source/components/layout/layout_task.rs#646), which impl's from a `LayoutTaskFactory` trait, whose `create` method relies on a `CompositorConstellationChan`.

I'm not sure whether I'm supposed to change the entire trait, and change those layout task related stuff. Any suggestions?

(oh, and sorry to smack you with this on your return from vacation :P)

jdm commented:
I suspect layout could move to using a ScriptConstellationChan (which might be more aptly named ContentConstellationChan, in that case)?

jdm commented:
So yeah, I would try modifying the trait to use ScriptConstellationChan, which will require moving the StylesheetLoadComplete, ReflowComplete, and TickAllAnimations messages from ConstellationControlMsg into MsgFromScript.

Wafflespeanut commented:
(sigh), forgot the recently added tidy checks

jdm commented:
As I feared, we have a lot of code moving around here and we still can't achieve my original goal of making `compositing` depend entirely on `gfx_traits` and not `gfx`. This is because of the `LayoutToPaint` msg that is in `gfx` that `compositing` relies upon.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7401) made this pull request unmergeable. Please resolve the merge conflicts.

Wafflespeanut commented:
@jdm Yep, you were right. But, isn't there any way (at all?) to break that dependency? Only the `LayoutToPaintMsg` is stopping us.

Commit:
Moving the shared ChromtToPaintMsg enum to gfx_traits; r=jdm

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7391) made this pull request unmergeable. Please resolve the merge conflicts.

Wafflespeanut commented:
@jdm I guess, we can't move only the `Msg` after all.

`LayoutToPaintMsg` (which is now in gfx_traits) demands `StackingContext`, whose impl requires PaintContext (which forces us to move the whole paint_context.rs) and then, PaintContext needs filters and `FontContext` (so, we gotta move font_context.rs) - now, `FontContext` needs font, font_cache_task, font_template & platform (and some of them demand text too).

Does this mean we'll be moving everything other than the remains of paint_task to gfx_traits? Am I complicating something, or is this the way it's always been? Because, I fear whether the future gfx crate would only have paint_task :)

jdm commented:
I haven't come up with any great solutions. Sorry! Thanks for settling the question of whether this was actually feasible, though!

Wafflespeanut commented:
@jdm No problem, I had fun time with it :)
